src/vs/workbench/parts/files/electron-browser/views/explorerViewer.ts

```diff
@@ -949,43 +949,49 @@ export class FileDragAndDrop extends SimpleFileResourceDragAndDrop {
 
 			return updateConfirmSettingsPromise.then(() => {
 				if (res.confirmed) {
-					return TPromise.join(sources.map(source => this.doHandleExplorerDrop(tree, source, target, isCopy))).then(() => void 0);
+					const rootDropPromise = this.doHandleRootDrop(sources.filter(s => s.isRoot), target);
+					return TPromise.join(sources.filter(s => !s.isRoot).map(source => this.doHandleExplorerDrop(tree, source, target, isCopy)).concat(rootDropPromise)).then(() => void 0);
 				}
 
 				return TPromise.as(void 0);
 			});
 		});
 	}
 
-	private doHandleExplorerDrop(tree: ITree, source: FileStat, target: FileStat | Model, isCopy: boolean): TPromise<void> {
-		return tree.expand(target).then(() => {
-			if (source.isRoot) {
-				const folders = this.contextService.getWorkspace().folders;
-				let sourceIndex: number;
-				let targetIndex: number;
-				const workspaceCreationData: IWorkspaceFolderCreationData[] = [];
-				const targetUri = target instanceof FileStat ? target.resource : folders[folders.length - 1].uri;
-
-				for (let index = 0; index < folders.length; index++) {
-					if (folders[index].uri.toString() === source.resource.toString()) {
-						sourceIndex = index;
-					}
-					if (folders[index].uri.toString() === targetUri.toString()) {
-						targetIndex = index;
-					}
-					workspaceCreationData.push({
-						name: folders[index].name,
-						uri: folders[index].uri
-					});
-				}
+	private doHandleRootDrop(roots: FileStat[], target: FileStat | Model): TPromise<void> {
```

Comment:
Wouldn't it be much easier here to do something like:
* fill an array of roots that includes everything except the roots that are being dragged
* find the index where to drop to (if target is model, the index is the length of the array)
* add the roots that are being dragged into the array at the correct location
* call workspace editing service with that array

Reply:
Thanks for the suggestion. This simplifies a bit. This was tackled via 5564db9710